### PR TITLE
Release mechanism - every push will trigger a release

### PR DESCRIPTION
**Pull request by madsop-nav**
Push to anything but main will trigger a dev release, every push to main a proper GitHub release.

Not sure if you've discussed the release process yet?

Might be worth looking into [JReleaser](https://jreleaser.org/)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -51,6 +51,11 @@
       <id>sonatype-nexus-staging</id>
       <url>https://s01.oss.sonatype.org/service/local/staging/deploy/maven2</url>
     </repository>
+    <repository>
+      <id>github</id>
+      <name>GitHub Packages</name>
+      <url>https://maven.pkg.github.com/madsop-nav/openhtmltopdf</url>
+    </repository>
```
Comment (siegelzc) [inline pom.xml:58]: Maven only allows a single <repository> to be specified.

Comment (siegelzc) [reply pom.xml:58]: We can use a profile block to override the repository when a github deployment flag is set.
This post describes the general pattern: https://stackoverflow.com/questions/16340673/setting-up-maven-to-allow-deployment-to-different-repositories-easily

Comment (aalmiray) [reply pom.xml:58]: JReleaser supports deploying to multiple locations during a single execution 😏

**Comment by madsop-nav** [comment]
After a lot of back and forth, this PR now releases properly - see examples at my profile, https://github.com/madsop-nav/openhtmltopdf/actions/runs/7902167607

However, the run at this repo fails with 403: `Caused by: org.apache.maven.shared.transfer.artifact.deploy.ArtifactDeployerException: Failed to deploy artifacts: Could not transfer artifact at.datenwort.openhtmltopdf:openhtmltopdf-parent:pom:1.1.5-20240214.135040-1 from/to github (https://maven.pkg.github.com/openhtmltopdf/openhtmltopdf): authorization failed for https://maven.pkg.github.com/openhtmltopdf/openhtmltopdf/at/datenwort/openhtmltopdf/openhtmltopdf-parent/1.1.5-SNAPSHOT/openhtmltopdf-parent-1.1.5-20240214.135040-1.pom, status: 403 Forbidden`

Might be something with the PR coming from a fork, I'm not sure..

File: pom.xml [hunk 2]
```diff
@@ -74,9 +63,47 @@
...
+      <activation>
+        <property>
+          <name>github-release</name>
+        </property>
+      </activation>
+      <distributionManagement>
+        <repository>
+          <id>github</id>
+          <name>GitHub Packages</name>
+          <url>https://maven.pkg.github.com/${env.GITHUB_REPOSITORY}</url>
```
Comment (siegelzc) [inline pom.xml:85]: Where is the environment variable being set? Is this built in to GitHub Actions?

Comment (madsop-nav) [reply pom.xml:85]: Yep, it is. https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables
image

**Comment by siegelzc** [comment]
The 403 error shown in "Create release -dev", we think this is due to the missing `GITHUB_TOKEN`?

So since it only has read access, you can build the code, but not deploy the release to the github package management server.

If this^ understanding is correct, I'm fine merging it, assuming you don't have anything else you want to do first.

**Comment by siegelzc** [comment]
Also, do you want me to add you to the GitHub organization?

**Comment by madsop-nav** [comment]
> The 403 error shown in "Create release -dev", we think this is due to the missing `GITHUB_TOKEN`?
> 
> So since it only has read access, you can build the code, but not deploy the release to the github package management server.
> 
> If this^ understanding is correct, I'm fine merging it, assuming you don't have anything else you want to do first.

Yes, that's precisely what I believe is happening. As the release is fully functional on my fork, and there are hardly any other differences between this repo and my fork, and it fits with the explanation from the Github docs, it's the far most likely reason.

I don't have anything else I'm intending to do in this PR, so let's indeed ship it.

**Comment by madsop-nav** [comment]
> Also, do you want me to add you to the GitHub organization?

That would be nice! My team is using this library every day, so I'm happy to contribute a bit.

**Comment by ganomi** [comment]
Does this mean the project should follow a gitflow like approach with a development branch that gets merged manually to main for an explicit release? And developers should open merge requests against development branch?

Or should developers still open merge requests against main and each feature merge will result in a release?

**Comment by madsop-nav** [comment]
> Does this mean the project should follow a gitflow like approach with a development branch that gets merged manually to main for an explicit release? And developers should open merge requests against development branch?
> 
> Or should developers still open merge requests against main and each feature merge will result in a proper release?
> 
> (Personally i prefer the gitflow approach.)

This PR is not intended to be opinionated in any sense. I'm not sure if the release process has been discussed since the transition to the new organization, but it sounds like that's something to discuss indeed. I'll create an issue for further discussion on that topic.

**Comment by madsop-nav** [comment]
> The 403 error shown in "Create release -dev", we think this is due to the missing `GITHUB_TOKEN`?
> 
> So since it only has read access, you can build the code, but not deploy the release to the github package management server.
> 
> If this^ understanding is correct, I'm fine merging it, assuming you don't have anything else you want to do first.

@siegelzc What do you think about merging this one now?

There is of course a risk of my 403 theory being wrong, and that it's something else causing the error, but I think the best way to figure out is to try.

**Review by siegelzc — APPROVED** [review]

**Comment by siegelzc** [comment]
> > The 403 error shown in "Create release -dev", we think this is due to the missing `GITHUB_TOKEN`?
> > So since it only has read access, you can build the code, but not deploy the release to the github package management server.
> > If this^ understanding is correct, I'm fine merging it, assuming you don't have anything else you want to do first.
> 
> @siegelzc What do you think about merging this one now?
> 
> There is of course a risk of my 403 theory being wrong, and that it's something else causing the error, but I think the best way to figure out is to try.

Go for it

**Comment by madsop-nav** [comment]
> > > The 403 error shown in "Create release -dev", we think this is due to the missing `GITHUB_TOKEN`?
> > > So since it only has read access, you can build the code, but not deploy the release to the github package management server.
> > > If this^ understanding is correct, I'm fine merging it, assuming you don't have anything else you want to do first.
> > 
> > 
> > @siegelzc What do you think about merging this one now?
> > There is of course a risk of my 403 theory being wrong, and that it's something else causing the error, but I think the best way to figure out is to try.
> 
> Go for it

Nice. Looks like I'm not allowed to merge, though. Could you help me out either by giving me write access or merging this PR?
image

**Comment by siegelzc** [comment]
You have write access now. Sorry, I thought making you a member of the organization would give you that permission implicitly.

**Pull request merged**